File: sokoban/map.py
from state import State

class Map:
    def __init__(self, name, width, height, fields, player_position, crate_positions):
        self._check_size(width, height, fields)
        self.name = name
        self.width = width
        self.height = height
        self.fields = fields
        self._initial_player_position = player_position
        self._initial_crate_positions = crate_positions

    def _check_size(self, width, height, fields):
        if len(fields) != width * height:
            raise ValueError()

    def __getitem__(self, coordinates):
        x, y = coordinates
        if x < 0 or y < 0 or x >= self.width or y >= self.height:
            raise ValueError()
        return self.fields[y * self.width + x]

    def accessible(self, x, y):
        if x < 0 or y < 0 or x >= self.width or y >= self.height:
            return False
        else:
            return self.fields[y * self.width + x].accessible

    def create_initial_state(self):
        return State(self._initial_player_position, self._initial_crate_positions)

File: web-sokoban/map.js
import {State} from './state.js'

export class Map {
    constructor(name, width, height, fields, player_position, crate_positions) {
        this._check_size(width, height, fields)
        this.name = name
        this.width = width
        this.height = height
        this.fields = fields
        this._initial_player_position = player_position
        this._initial_crate_positions = crate_positions
    }

    _check_size(width, height, fields) {
        if (len(fields) !== width * height) {
            throw new ValueError()
        }
    }

    __getitem__(coordinates) {
        if (this._outOfBounds(coordinates.x, coordinates.y)) {
            throw new ValueError()
        }
        return this.fields[y * this.width + x]
    }

    _outOfBounds(x, y) {
        return x < 0 || y < 0 || x >= this.width || y >= this.height
    }

    accessible(x, y) {
        if (x < 0 || y < 0 || x >= this.width || y >= this.height) {
            return false
        } else {
            return this.fields[y * this.width + x].accessible
        }
    }

    create_initial_state() {
        return State(this._initial_player_position, this._initial_crate_positions)
    }
}
